src/main/java/com/google/step/snippet/servlets/UserSettingServlet.java

```diff
@@ -24,11 +29,38 @@
 /** Servelet that renders user setting page */
 @WebServlet("/user")
 public class UserSettingServlet extends HttpServlet {
+  private static final String WEBSITE_PARAMETER = "website";
+  private static final String LANGUAGE_PARAMETER = "language";
+  private static final String ID_PARAMETER = "id";
+  private static final String USER_PARAMETER = "UserInfo";
+  private String referer = "";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
     // Forward the request to the template (which is a servlet itself).
     request.getRequestDispatcher("WEB-INF/templates/user_dashboard.jsp").forward(request, response);
+    referer = request.getHeader("referer");
```

Comment:
I think you meant to add this in the doPost in place of line 50 `response.sendRedirect("/home");` rather than in the doGet call here? Or maybe I'm just misunderstanding what the doGet function is intended to do now or eventually

Reply:
the referer of doPost method is /user, doPost happens when you click the submit button on /user page to update information about user

Reply:
ah ok, that makes sense for wanting to have doPost to always redirect to /home.

however, I'm not sure what the doGet function does here though / how it will be used? It seems unnecessary.

Reply:
doGet just direct user to the setting page

Reply:
Please do not read the "referer" header. You cannot trust it. Anyone can modify it. Browsers sometimes strip it out for privacy reasons. It is not a reliable way at all to check the referring page.

Reply:
(Also, this referer field is modified by _any_ request. User B can change the redirect for User A by loading this page after User A does.)

May I suggest that POST requests load this page again? (Assuming that the UI shows your current settings.)

Reply:
so now POST reload the page with new settings, and a button redirect to homepage is added